### PR TITLE
[BugFix]Support clone expression on grouping set

### LINKED ISSUE #5419: result error when use aggregate functions with rollup
### Steps to reproduce the behavior (Required)
```
CREATE TABLE IF NOT EXISTS t1 (
                     id INT,
                     v1 INT
                 )
                 DUPLICATE KEY(id)
                 DISTRIBUTED BY HASH(id) BUCKETS 3
                 PROPERTIES ("replication_num" = "1");
insert into t1 values (1,2);
select v1,sum(id) from t1 group by rollup(v1, id) order by v1;
```

### Expected behavior (Required)
```
+------+---------+
| v1   | sum(id) |
+------+---------+
| NULL |    1|
|    2 |       1 |
|    2 |    1|
+------+---------+
```
### Real behavior (Required)
```
+------+---------+
| v1   | sum(id) |
+------+---------+
| NULL |    NULL |
|    2 |       1 |
|    2 |    NULL |
+------+---------+
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------------------------+
| current_version()                        |
+------------------------------------------+
| PIPELINE_SQLANCER_MASTER_RELEASE 4bba40d |
+------------------------------------------+
```

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5419

## Problem Summary(Required) ：

```
select v1,sum(id) from t1 group by rollup(v1, id) order by v1;
```

the currently plan:
```
MySQL n0> explain select v1,sum(id) from t1 group by rollup(v1, id) order by v1;

|   2:AGGREGATE (update serialize)                            |
|   |  STREAMING                                              |
|   |  output: sum(1: id)                                     |
|   |  group by: 2: v1, 1: id, 4: GROUPING_ID                 |
|   |                                                         |
|   1:REPEAT_NODE                                             |
|   |  repeat: repeat 2 lines [[], [2], [1, 2]]               |
|   |                                                         |
|   0:OlapScanNode                                            |
```

Aggregate function use `id` column which one was add `null` value, not origin column. So we add clone expr to copy origin column before RepeatNode

after plan:
```
|   3:AGGREGATE (update serialize)                            |
|   |  STREAMING                                              |
|   |  output: sum(3: expr)                                   |
|   |  group by: 2: v1, 1: id, 5: GROUPING_ID                 |
|   |                                                         |
|   2:REPEAT_NODE                                             |
|   |  repeat: repeat 2 lines [[], [2], [1, 2]]               |
|   |                                                         |
|   1:Project                                                 |
|   |  <slot 1> : 1: id                                       |
|   |  <slot 2> : 2: v1                                       |
|   |  <slot 3> : clone(1: id)                                |
|   |                                                         |
|   0:OlapScanNode                                            |  
```

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 40 / 48 (83.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/AggregationAnalyzer.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CloneOperator.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorsReuse.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|14|14|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 40 / 48 (83.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/AggregationAnalyzer.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CloneOperator.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorsReuse.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|14|14|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|1|1|100.00%|

File: be/src/exprs/vectorized/clone_expr.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2022 StarRocks Limited.
+
+#pragma once
+
+#include "exprs/expr.h"
+
+namespace starrocks::vectorized {
+
+class VectorizedCloneExpr final : public Expr {
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/clone_expr.h:9]: CloneExpr is enough?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java [hunk 2]
```diff
@@ -288,13 +289,26 @@ private OptExprBuilder limit(OptExprBuilder subOpt, LimitElement limit) {
...
         }
 
+        // handle aggregate function use grouping set columns, use clone column replace original column
+        // e.g:
+        // before: select sum(a) from xx group by rollup(a);
+        // after: select sum(clone(a)) from xx group by rollup(a);
+        if (groupingSetsList != null) {
+            for (Expr groupBy : groupByExpressions) {
+                aggregates.forEach(f -> f.getParams().exprs()
+                        .replaceAll(root -> replaceExprBottomUp(root, groupBy, new CloneExpr(groupBy))));
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java:306]: Could use visitor like NotFullGroupByRewriter

File: be/src/exprs/vectorized/clone_expr.h (hunk 1 above)
Comment (Seaven) [reply be/src/exprs/vectorized/clone_expr.h:9]: Enough

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java:306]: It's different, `NotFullGroupByRewriter` only reslove `SlotExpr`, we need reslove function. Like:
```
select v1,sum(id + 1) from t1 group by rollup(v1, id + 1) order by v1;
```

**Comment by HangyuanLiu** [comment]
Why not only change the output logic of repeatNode, but add a cloneExpr separately, this is not a logic that many codes need to be reused, and it increases the difficulty for users to understand the logic plan

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java [hunk 3]
```diff
@@ -288,13 +289,26 @@ private OptExprBuilder limit(OptExprBuilder subOpt, LimitElement limit) {
...
+                aggregates.replaceAll(
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java:307]: Why replace aggreage function? plz give a example

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java:307]: look Problem Summary(Required)

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 40 / 48 (83.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/AggregationAnalyzer.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CloneOperator.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorsReuse.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|14|14|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|1|1|100.00%|

**Review by kangkaisen — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**